File: deepseek-from-scratch-python/src/deepseek/training/grpo.py
import torch
import torch.nn as nn
import torch.nn.functional as F

class GRPOTrainer:
    def __init__(self, beta=0.01):
        self.beta = beta
        
    def compute_loss(self, logits, input_ids, rewards, ref_logits):
        # logits: (G, Seq, Vocab)
        # input_ids: (G, Seq)
        # rewards: (G,)
        # ref_logits: (G, Seq, Vocab)
        
        G, Seq, _ = logits.shape
        
        # 1. Advantages
        mean_r = rewards.mean()
        std_r = rewards.std() + 1e-8
        advantages = (rewards - mean_r) / std_r # (G,)
        
        # 2. Policy Log Probs
        log_probs = F.log_softmax(logits, dim=-1)
        # Gather log probs of actual tokens
        # input_ids: (G, Seq) -> (G, Seq, 1)
        token_log_probs = log_probs.gather(2, input_ids.unsqueeze(-1)).squeeze(-1) # (G, Seq)
        seq_log_probs = token_log_probs.sum(dim=1) # (G,)
        
        # 3. KL Divergence
        ref_log_probs = F.log_softmax(ref_logits, dim=-1)
        # KL(P || Ref) = sum(P * (log P - log Ref))
        # Approx: log P - log Ref (token level)
        # Wait, exact KL is sum P * (log P - log Ref)
        probs = F.softmax(logits, dim=-1)
        kl = (probs * (log_probs - ref_log_probs)).sum(dim=-1) # (G, Seq)
        mean_kl = kl.mean(dim=1) # (G,)
        
        # 4. Loss
        # Loss = - (Adv * seq_log_probs) + beta * mean_kl
        loss = - (advantages * seq_log_probs) + self.beta * mean_kl
        return loss.mean()

class GroupSampler:
    def __init__(self, group_size):
        self.group_size = group_size
        
    def sample(self, prompt):
        return [f"Sampled output {i} for prompt..." for i in range(self.group_size)]

File: Deepseek-from-scratch-in-rust/src/training/grpo.rs
use candle_core::{Result, Tensor};

// GRPO (Group Relative Policy Optimization) Simulation
// GRPO removes the critic (value function) and uses group-relative rewards as the baseline.
// Loss = -1/G * Sum( A_i * log P(o_i | q) ) + beta * KL(P || Ref)
// Where A_i = (r_i - mean(r)) / std(r)

pub struct GRPOTrainer {
    // In a real scenario, this would hold the policy model and reference model.
    // Here we simulate the loss calculation logic.
    beta: f64, // KL penalty coefficient
}

impl GRPOTrainer {
    pub fn new(beta: f64) -> Self {
        Self { beta }
    }

    // Simulate computing the GRPO loss for a group of outputs
    // logits: (G, Seq, Vocab) - Logits for the group of G outputs
    // input_ids: (G, Seq) - Token IDs for the group
    // rewards: (G,) - Scalar reward for each output in the group
    // ref_logits: (G, Seq, Vocab) - Logits from the reference model (for KL)
    pub fn compute_loss(
        &self,
        logits: &Tensor,
        input_ids: &Tensor,
        rewards: &Tensor,
        ref_logits: &Tensor,
    ) -> Result<Tensor> {
        let (g, seq, _) = logits.dims3()?;
        
        // 1. Compute Advantages: A_i = (r_i - mean(r)) / std(r)
        let mean_r = (rewards.sum_all()? / g as f64)?;
        let mean_r_bcast = mean_r.broadcast_as(rewards.shape())?;
        
        // Variance: sum((r - mean)^2) / G
        let diff = (rewards - &mean_r_bcast)?;
        let var = (diff.sqr()?.sum_all()? / g as f64)?;
        let std = (var.sqrt()? + 1e-8)?; // Add epsilon
        let std_bcast = std.broadcast_as(rewards.shape())?;
        
        let advantages = (diff / std_bcast)?; // (G,)
        
        // 2. Compute Policy Log Probs: log P(o_i | q)
        // Gather log probs of the actual tokens
        // logits -> log_softmax -> gather(input_ids)
        let log_probs = candle_nn::ops::log_softmax(logits, 2)?;
        let log_probs_tokens = log_probs.gather(&input_ids.unsqueeze(2)?, 2)?.squeeze(2)?; // (G, Seq)
        
        // Sum log probs over sequence (assuming full sequence is generated)
        let seq_log_probs = log_probs_tokens.sum(1)?; // (G,)
        
        // 3. Compute KL Divergence: KL(P || Ref)
        // KL = sum(P * (log P - log Ref))
        // Approx: log P - log Ref (token level)
        let ref_log_probs = candle_nn::ops::log_softmax(ref_logits, 2)?;
        let kl = (log_probs.exp()? * (log_probs - ref_log_probs)?)?.sum(2)?; // (G, Seq)
        let mean_kl = (kl.sum(1)? / seq as f64)?; // (G,) - mean KL per token or sum? Usually sum or mean. Let's use mean per token.
        
        // 4. Final Loss
        // Loss = - (Advantages * seq_log_probs) + beta * mean_kl
        // We average over the group G
        
        let adv_loss = (advantages * seq_log_probs)?;
        let kl_penalty = (mean_kl * self.beta)?;
        
        let loss = (kl_penalty - adv_loss)?;
        let mean_loss = (loss.sum_all()? / g as f64)?;
        
        Ok(mean_loss)
    }
}

// Group Sampler Simulation
pub struct GroupSampler {
    group_size: usize,
}

impl GroupSampler {
    pub fn new(group_size: usize) -> Self {
        Self { group_size }
    }
    
    // Simulate sampling G outputs for a single prompt
    pub fn sample(&self, _prompt: &str) -> Result<Vec<String>> {
        let mut outputs = Vec::new();
        for i in 0..self.group_size {
            outputs.push(format!("Sampled output {} for the prompt...", i));
        }
        Ok(outputs)
    }
}
